### ISSUE TITLE
Fix S2589 FP: When delegates stored in members are called

### ISSUE THREAD

**Issue by bitbonk** [SonarSource/sonar-dotnet#9746]
### Description

Consider the code below.

Obviously `this.OnNext` stores a delegate that is invoked between the two `this.IsPaused` calls. That invocation could potentially have changed `this.IsPaused`.

### Reproducer

```csharp
if (this.IsPaused)
{
    return;
}

if (this.OnNext is not null)
{
    QueuedOperation nextOperation = this.pendingOperations.Peek();

    if (!nextOperation.IsCancellationRequested)
    {
        // This operation is not yet canceled and did not yet time out, call OnNext for it.
        this.OnNext(nextOperation.Argument, nextOperation.CancellationToken);

        // check if OnNext paused the queue
        if (this.IsPaused) // false positive for S2589
        { 
            return;
        }
    }
}
```

### Product and Version

SonarQube Server, Community Build v25.4.0.105899 Standard Experience

**Comment by mary-georgiou** [comment]
Hello @bitbonk and thanks for reporting this.

If I understood correctly - when you call `this.OnNext(nextOperation.Argument, nextOperation.CancellationToken);` there's an effect on `this.IsPaused` due to the method/methods that might be called behind the delegate right?


Also one question - could you please recheck the product + version? This rule is not shipped with SQ Community build.
Thanks!

**Comment by bitbonk** [comment]
You understood correctly. I'll only be able to check the version in a few weeks from now.

**Comment by mary-georgiou** [comment]
Ok no worries - since I can repro with the latest analyzer version means that the issue is still there.

I confirm this is an FP and it's due to the fact that we don't yet manage delegates. 
I added an issue to our backlog https://sonarsource.atlassian.net/browse/NET-2157 (internal only) - we'll deal with it in a future sprint.
Thanks!

**Comment by martin-strecker-sonarsource** [comment]
Hello @bitbonk,

From your code, it isn't clear how `OnNext` is able to set the IsPaused state. I assume the class is defined about like this:
```cs
public class OperationExecuter
{
    protected Action<object, CancellationToken> OnNext { get; set; }
    public bool IsPaused { get; private set; }

    protected virtual void Init()
    {
        this.OnNext = (a, c) => { this.IsPaused = true; };
    }

    void Execute()
    {
        if (this.IsPaused)
        {
            return;
        }

        if (this.OnNext is not null)
        {
            // Simplified code here:
            this.OnNext(null, CancellationToken.None);

            if (this.IsPaused) // false positive for S2589
            { 
                return;
            }
        }
    }
}
```

If this is about correct, I'm curious to understand why the `OnNext` property is defined like that. I assume `IsPaused` has a private or protected accessiblity when it comes to setting the field/property, right? If so, the `OnNext` delegate can only bet set from within the same class or derived class, because otherwise, it would not be possible to change `IsPaused` inside the delegate, because
* `this` is not in scope and
* `this.IsPaused` is not accessible.

If that is the case and `OnNext` can only be set from within the class, I wonder why `OnNext` isn't just a virtual method here. Why do you need it to be a delegate field/property?

Background: We would not raise if `OnNext` would be an instance method of that class, because we would assume it can reset the state of `IsPaused`. But if `OnNext` is a delegate, we assume the delegate is set from the outside (e.g. via the constructor: `new OperationExecuter((argument, cancellationToken) => { ... }`). If the delegate is passed from outside the class `IsPaused` can not be set.
I'm a bit hesitant to also consider the case with the delegate invocation on `OnNext` to also be able to reset `IsPaused` because I would argue that `OnNext` should be a virtual method instead of a delegate field.

Can you please explain, why `OnNext` is a delegate field and not a virtual method? Maybe I miss something important here.

**Comment by bitbonk** [comment]
The class has public `Pause()` and `Resume()` instance method that simply set the `IsPaused` property. `OnNext` has a public setter so users can "register" their own delegate. So within the `OnNext` delegate user code can call `Pause()` and `Resume()` and thereby changing the `IsPaused`property.

**Comment by martin-strecker-sonarsource** [comment]
Thank you for the explanation. So `IsPaused` isn't really private but can change by a call to public methods. That looks indeed like a pattern that is common enough. We will do the following: We will consider the `OnNext()` invocation being like a normal instance method call and therefore assume that it might change private state. We will then take a look how many issues we loose in our test suite (~60 open source projects). I assume that most of these missing issues are actually real issues. If the number of lost issues is too high, we will roll back the change and document this pattern as a known false positive (we raise an issue, but there isn't a problem).

**Comment by martin-strecker-sonarsource** [comment]
Fixed in version [10.15.0.120848](https://github.com/SonarSource/sonar-dotnet/releases/tag/10.15.0.120848)
